File: Examples/DataManagementFileExternalMemory/src/course/examples/datamanagement/fileexternal/ExternalFileWriteReadActivity.java
package course.examples.datamanagement.fileexternal;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import android.app.Activity;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.widget.ImageView;

public class ExternalFileWriteReadActivity extends Activity {
	private final String fileName = "painter.png";
	private String TAG = "ExternalFileWriteReadActivity";

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.main);

		if (Environment.MEDIA_MOUNTED.equals(Environment
				.getExternalStorageState())) {

			File outFile = new File(
					getExternalFilesDir(Environment.DIRECTORY_PICTURES),
					fileName);
			
			if (!outFile.exists())
				copyImageToMemory(outFile);
			
			ImageView imageview = (ImageView) findViewById(R.id.image);
			imageview.setImageURI(Uri.parse("file://" + outFile.getAbsolutePath()));
		
		}
	}

	private void copyImageToMemory(File outFile) {
		try {

			BufferedOutputStream os = new BufferedOutputStream(
					new FileOutputStream(outFile));

			BufferedInputStream is = new BufferedInputStream(getResources()
					.openRawResource(R.raw.painter));

			copy(is, os);

		} catch (FileNotFoundException e) {
			Log.e(TAG, "FileNotFoundException");
		}
	}

	private void copy(InputStream is, OutputStream os) {
		final byte[] buf = new byte[1024];
		int numBytes;
		try {
			while (-1 != (numBytes = is.read(buf))) {
				os.write(buf, 0, numBytes);
			}
		} catch (IOException e) {
			e.printStackTrace();
		} finally {
			try {
				is.close();
				os.close();
			} catch (IOException e) {
				Log.e(TAG, "IOException");

			}
		}
	}
}

File: ExamplesKotlin/DataManagementFileExternalMemory/app/src/main/java/course/examples/datamanagement/fileexternal/ExternalFileWriteReadActivity.kt
package course.examples.datamanagement.fileexternal

import android.app.Activity
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.util.Log
import android.widget.ImageView
import java.io.*

class ExternalFileWriteReadActivity : Activity() {

    companion object {
        private const val TAG = "ExtFileWriteRead"
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main)

        if (Environment.MEDIA_MOUNTED == Environment
                .getExternalStorageState()
        ) {

            val fileName = "painter.png"
            val outFile = File(
                getExternalFilesDir(Environment.DIRECTORY_PICTURES), fileName)

            if (!outFile.exists())
                copyImageToMemory(outFile)

            val imageview = findViewById<ImageView>(R.id.image)
            imageview.setImageURI(Uri.parse("file://" + outFile.absolutePath))

        }
    }

    private fun copyImageToMemory(outFile: File) {

        var outputStream: BufferedOutputStream? = null
        var inputStream: BufferedInputStream? = null

        try {

            outputStream = BufferedOutputStream(
                FileOutputStream(outFile)
            )

            inputStream = BufferedInputStream(
                resources.openRawResource(R.raw.painter)
            )

            //copy(inputStream, outputStream)
            inputStream.copyTo(outputStream)

        } catch (e: FileNotFoundException) {
            Log.e(TAG, "FileNotFoundException")
        } finally {
            try {
                inputStream?.close()
                outputStream?.close()
            } catch (e: IOException) {
                Log.e(TAG, "IOException")
            }

        }
    }
}